### PR TITLE
Retry Pod/RC updates in kubectl rolling-update

### LINKED ISSUE #27328: Kubectl client Kubectl rolling-update should support rolling-update to same image [Conformance] {Kubernetes e2e suite}
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubernetes-e2e-gke-staging-parallel/4156/

Failed: Kubectl client Kubectl rolling-update should support rolling-update to same image [Conformance] {Kubernetes e2e suite}

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:1029
Expected error:
    <*errors.errorString | 0xc2084d0fc0>: {
        s: "Error running &{/workspace/kubernetes/platforms/linux/amd64/kubectl [kubectl --server=https://130.211.160.169 --kubeconfig=/workspace/.kube/config rolling-update e2e-test-nginx-rc --update-period=1s --image=gcr.io/google_containers/nginx:1.7.9 --image-pull-policy=IfNotPresent --namespace=e2e-tests-kubectl-nlrmc] []  <nil> Created e2e-test-nginx-rc-11e19e8c183ac3df74483689a98697dc\n Error from server: replicationcontrollers \"e2e-test-nginx-rc\" cannot be updated: the object has been modified; please apply your changes to the latest version and try again\n [] <nil> 0xc208308640 exit status 1 <nil> true [0xc208050ec0 0xc208050ee0 0xc208050f00] [0xc208050ec0 0xc208050ee0 0xc208050f00] [0xc208050ed8 0xc208050ef8] [0x968650 0x968650] 0xc2088a8660}:\nCommand stdout:\nCreated e2e-test-nginx-rc-11e19e8c183ac3df74483689a98697dc\n\nstderr:\nError from server: replicationcontrollers \"e2e-test-nginx-rc\" cannot be updated: the object has been modified; please apply your changes to the latest version and try again\n\nerror:\nexit status 1\n",
    }
    Error running &{/workspace/kubernetes/platforms/linux/amd64/kubectl [kubectl --server=https://130.211.160.169 --kubeconfig=/workspace/.kube/config rolling-update e2e-test-nginx-rc --update-period=1s --image=gcr.io/google_containers/nginx:1.7.9 --image-pull-policy=IfNotPresent --namespace=e2e-tests-kubectl-nlrmc] []  <nil> Created e2e-test-nginx-rc-11e19e8c183ac3df74483689a98697dc
     Error from server: replicationcontrollers "e2e-test-nginx-rc" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
     [] <nil> 0xc208308640 exit status 1 <nil> true [0xc208050ec0 0xc208050ee0 0xc208050f00] [0xc208050ec0 0xc208050ee0 0xc208050f00] [0xc208050ed8 0xc208050ef8] [0x968650 0x968650] 0xc2088a8660}:
    Command stdout:
    Created e2e-test-nginx-rc-11e19e8c183ac3df74483689a98697dc
    
    stderr:
    Error from server: replicationcontrollers "e2e-test-nginx-rc" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
    
    error:
    exit status 1
    
not to have occurred
```

### PR DESCRIPTION
**Pull request by janetkuo**
Fixes #27328

@kubernetes/kubectl 

[Analytics]()

### CODE REVIEW

File: pkg/kubectl/rolling_updater.go [hunk 1]
```diff
@@ -643,19 +643,26 @@ func SetNextControllerAnnotation(rc *api.ReplicationController, name string) {
...
+			return nil, err
+		}
+		return oldRc, err
 	}
 }
 
 const MaxRetries = 3
 
 func AddDeploymentKeyToReplicationController(oldRc *api.ReplicationController, client client.Interface, deploymentKey, deploymentValue, namespace string, out io.Writer) (*api.ReplicationController, error) {
+	fmt.Printf("calling AddDeploymentKeyToReplicationController\n")
```
Comment (kargakis) [inline pkg/kubectl/rolling_updater.go]: remove this?

File: pkg/kubectl/rolling_updater.go [hunk 2]
```diff
@@ -763,6 +759,33 @@ func updateWithRetries(rcClient client.ReplicationControllerInterface, rc *api.R
 	return rc, err
 }
 
+type updateFuncForPod func(pod *api.Pod)
+
+// updateWithRetriesForPod updates applies the given pod as an update.
+func updateWithRetriesForPod(podClient client.PodInterface, pod *api.Pod, applyUpdate updateFuncForPod) (*api.Pod, error) {
+	var err error
+	oldPod := pod
+	err = wait.Poll(10*time.Millisecond, 1*time.Minute, func() (bool, error) {
```
Comment (kargakis) [inline pkg/kubectl/rolling_updater.go]: 10ms is a short interval but it's fine if we backoff. Generally retries should backoff.

File: pkg/kubectl/rolling_updater.go [hunk 3]
```diff
@@ -763,6 +759,33 @@ func updateWithRetries(rcClient client.ReplicationControllerInterface, rc *api.R
 	return rc, err
 }
 
+type updateFuncForPod func(pod *api.Pod)
+
+// updateWithRetriesForPod updates applies the given pod as an update.
+func updateWithRetriesForPod(podClient client.PodInterface, pod *api.Pod, applyUpdate updateFuncForPod) (*api.Pod, error) {
```
Comment (kargakis) [inline pkg/kubectl/rolling_updater.go]: Since you are updating meta in both cases, you could have a generic `updateWithRetries` by using a meta accessor, no?

File: pkg/kubectl/rolling_updater.go (hunk 2 above)
Comment (pwittrock) [inline pkg/kubectl/rolling_updater.go]: Do you mean exponential backoff?  Or are you saying constant backoff is sufficient?

Comment (kargakis) [inline pkg/kubectl/rolling_updater.go]: Sorry for the ambiguity. Exponential backoff was what I was thinking.

File: pkg/kubectl/rolling_updater.go (hunk 3 above)
Comment (janetkuo) [inline pkg/kubectl/rolling_updater.go]: Not sure how since they're using different client interface?

File: pkg/kubectl/rolling_updater.go (hunk 1 above)
Comment (janetkuo) [inline pkg/kubectl/rolling_updater.go]: Done

File: pkg/kubectl/rolling_updater.go (hunk 2 above)
Comment (janetkuo) [inline pkg/kubectl/rolling_updater.go]: Done

File: pkg/kubectl/rolling_updater.go [hunk 4]
```diff
@@ -756,13 +749,53 @@ func updateWithRetries(rcClient client.ReplicationControllerInterface, rc *api.R
...
+		// The Get passed: pod contains the latest controller, expect a poll for the update.
+		return false
+	})
+	// If the error is non-nil the returned pod cannot be trusted, if it is nil, the returned
+	// pod contains the applied update.
+	return pod, err
+}
+
+// retry retries fn using exponential backoff
+func retry(delay time.Duration, maxRetries int, fn func() bool) error {
```
Comment (pwittrock) [inline pkg/kubectl/rolling_updater.go]: Take a look at https://github.com/kubernetes/kubernetes/blob/293b0d0815cf8f54ea1df802546b2f61ad4a2a01/pkg/client/unversioned/conditions.go#L67.  We probably want to copy this or move it into a util pkg.

File: pkg/kubectl/rolling_updater.go (hunk 3 above)
Comment (kargakis) [inline pkg/kubectl/rolling_updater.go]: uh, nvmd

File: pkg/kubectl/rolling_updater.go [hunk 5]
```diff
@@ -187,15 +187,15 @@ func (r *RollingUpdater) Update(config *RollingUpdaterConfig) error {
...
 			return err
 		}
-		if existing.Annotations == nil {
-			existing.Annotations = map[string]string{}
-		}
-		existing.Annotations[originalReplicasAnnotation] = strconv.Itoa(int(existing.Spec.Replicas))
-		updated, err := r.c.ReplicationControllers(existing.Namespace).Update(existing)
-		if err != nil {
+		originReplicas := strconv.Itoa(int(existing.Spec.Replicas))
+		if oldRc, err = updateWithRetries(r.c.ReplicationControllers(existing.Namespace), existing, func(rc *api.ReplicationController) {
```
Comment (pwittrock) [inline pkg/kubectl/rolling_updater.go]: pull the updateWithRetries out of the "if" condition as it makes it hard to follow

File: pkg/kubectl/rolling_updater.go (hunk 4 above)
Comment (AdoHe) [inline pkg/kubectl/rolling_updater.go]: Also we should make it clear under what scenario we should retry.

Comment (janetkuo) [inline pkg/kubectl/rolling_updater.go]: done

File: pkg/kubectl/rolling_updater.go (hunk 5 above)
Comment (janetkuo) [inline pkg/kubectl/rolling_updater.go]: done

**Comment by janetkuo** [comment]
@pwittrock @kargakis ptal

File: pkg/kubectl/rolling_updater.go [hunk 6]
```diff
@@ -736,33 +729,62 @@ func AddDeploymentKeyToReplicationController(oldRc *api.ReplicationController, c
...
 }
 
+type updatePodFunc func(controller *api.Pod)
+
+// updatePodWithRetries retries updating the given pod on conflict with the following steps:
+// 1. Get latest resource
+// 2. applyUpdate
+// 3. Update the resource
+func updatePodWithRetries(c client.Interface, namespace string, pod *api.Pod, applyUpdate updatePodFunc) (*api.Pod, error) {
+	oldPod := pod
```
Comment (kargakis) [inline pkg/kubectl/rolling_updater.go]: I think you need to deep-copy here

Comment (janetkuo) [inline pkg/kubectl/rolling_updater.go]: Fixed

**Comment by kargakis** [comment]
@janetkuo squash and reapply lgtm

**Comment by pwittrock** [comment]
@kargakis Thanks for doing the review

**Comment by erictune** [comment]
@janetkuo I think this is already in release-1.3 branch since it merged several days ago.  I'm removing the cherrypick-candidate label.  Re-add if i am mistaken.

**Comment by roberthbailey** [comment]
@erictune - fyi the label was also there because we wanted to cherry pick this into 1.2 (which was done yesterday). It's fine to remove it as this point.
